Sources/SKTestSupport/SkipUnless.swift

```diff
@@ -0,0 +1,252 @@
+          skipMessage: """
+            Skipping because toolchain has Swift version \(toolchainSwiftVersion) \
+            but test requires at least \(requiredSwiftVersion)
+            """
+        )
+      } else if toolchainSwiftVersion == requiredSwiftVersion {
+        logger.info("Checking if feature '\(featureName)' is supported")
+        if try await !featureCheck() {
+          checkResult = .featureUnsupported(skipMessage: "Skipping because toolchain doesn't contain \(featureName)")
+        } else {
+          checkResult = .featureSupported
+        }
+        logger.info("Done checking if feature '\(featureName)' is supported")
+      } else {
+        checkResult = .featureSupported
+      }
+    }
+    checkCache[featureName] = checkResult
+
+    if case .featureUnsupported(let skipMessage) = checkResult {
+      throw XCTSkip(skipMessage, file: file, line: line)
+    }
+  }
+
+  public static func sourcekitdHasSemanticTokensRequest(
+    file: StaticString = #file,
+    line: UInt = #line
+  ) async throws {
+    try await skipUnlessSupportedByToolchain(
+      swiftVersion: SwiftVersion(5, 11),
+      featureName: "semantic token support in sourcekitd",
+      file: file,
+      line: line
+    ) {
+      let testClient = try await TestSourceKitLSPClient()
+      let uri = DocumentURI.for(.swift)
+      testClient.openDocument("func test() {}", uri: uri)
+      do {
+        _ = try await testClient.send(DocumentSemanticTokensRequest(textDocument: TextDocumentIdentifier(uri)))
+      } catch let error as ResponseError {
+        print(error.message)
+        return !error.message.contains("unknown request: source.request.semantic_tokens")
+      }
+      return true
+    }
+  }
+
+  public static func sourcekitdSupportsRename(
+    file: StaticString = #file,
+    line: UInt = #line
+  ) async throws {
+    try await skipUnlessSupportedByToolchain(
+      swiftVersion: SwiftVersion(5, 11),
+      featureName: "rename support in sourcekitd",
+      file: file,
+      line: line
+    ) {
+      let testClient = try await TestSourceKitLSPClient()
+      let uri = DocumentURI.for(.swift)
+      let positions = testClient.openDocument("void 1️⃣test() {}", uri: uri)
+      do {
+        _ = try await testClient.send(
+          RenameRequest(textDocument: TextDocumentIdentifier(uri), position: positions["1️⃣"], newName: "test2")
+        )
+      } catch let error as ResponseError {
+        return error.message != "Running sourcekit-lsp with a version of sourcekitd that does not support rename"
+      }
+      return true
+    }
+  }
+
+  /// SwiftPM moved the location where it stores Swift modules to a subdirectory in
+  /// https://github.com/apple/swift-package-manager/pull/7103.
+  ///
+  /// sourcekit-lsp uses the built-in SwiftPM to synthesize compiler arguments and cross-module tests fail if the host
+  /// toolchain’s SwiftPM stores the Swift modules on the top level but we sythesize compiler arguments expecting the
```

Comment:
```suggestion
  /// toolchain’s SwiftPM stores the Swift modules on the top level but we synthesize compiler arguments expecting the
```

Having said that, can we not just check this? ie. check for a `Modules` folder and otherwise assume top level?

Reply:
Like, if SwiftPM is called from SourceKit-LSP, check whether the build folder looks like it was generated by a previous SwiftPM version and if so, generate the old-style compiler arguments?

CC @neonichu

Reply:
Sorry, what is generating the arguments here? By synthesizing I assumed that meant LSP, in which case, can't we just check if the folder exists? Or is it SwiftPM generating them?

Reply:
SwiftPM is generating them.

Reply:
Ah I see, `synthesize` just meant getting them. In that case... 💩 this is probably good enough then, suppose it's only for people on old toolchains anyway.

Reply:
> ```suggestion
> 
>   /// toolchain’s SwiftPM stores the Swift modules on the top level but we synthesize compiler arguments expecting the
> 
> ```
> 
> 
> 
> Having said that, can we not just check this? ie. check for a `Modules` folder and otherwise assume top level?